src/shared/previewUtils.ts

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+// **********************************************************************************************
+// * TODO: When we finalize the implementation for the preview commands and things settle down, *
+// *       consider moving most of these into PreviewUtils of lwc-dev-mobile-core instead.      *
+// **********************************************************************************************
+
+import fs from 'node:fs';
+import https from 'node:https';
+import os from 'node:os';
+import path from 'node:path';
+import { Logger, Messages } from '@salesforce/core';
+import {
+  AndroidAppPreviewConfig,
+  AndroidUtils,
+  AndroidVirtualDevice,
+  CommonUtils,
+  CryptoUtils,
+  IOSAppPreviewConfig,
+  IOSSimulatorDevice,
+  IOSUtils,
+  LaunchArgument,
+  PreviewUtils as LwcDevMobileCorePreviewUtils,
+  Platform,
+} from '@salesforce/lwc-dev-mobile-core';
+import { Progress, Spinner } from '@salesforce/sf-plugins-core';
+import fetch from 'node-fetch';
+
+Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
+const messages = Messages.loadMessages('@salesforce/plugin-lightning-dev', 'lightning.preview.app');
+const DevPreviewAuraMode = 'DEVPREVIEW';
+
+export class PreviewUtils {
+  public static generateWebSocketUrlForLocalDevServer(platform: string, port: string): string {
+    return LwcDevMobileCorePreviewUtils.generateWebSocketUrlForLocalDevServer(platform, port);
+  }
+
+  /**
+   * Attempts to fetch the targeted mobile device for previewing.
+   *
+   * @param platform A mobile platform (iOS or Android)
+   * @param deviceId An optional device identifier (such as name or UDID)
+   * @param logger An optional logger to be used for logging
+   * @returns The iOS or Android device, or `undefined` if not found
+   */
+  public static async getMobileDevice(
+    platform: Platform.ios | Platform.android,
+    deviceId?: string,
+    logger?: Logger
+  ): Promise<IOSSimulatorDevice | AndroidVirtualDevice | undefined> {
+    let device: IOSSimulatorDevice | AndroidVirtualDevice | undefined;
+
+    logger?.debug(`Attempting to get mobile device for platform ${platform}`);
+
+    if (deviceId) {
+      logger?.debug(`Attempting to get device ${deviceId}`);
+      device =
```

Comment:
It's not my favorite thing to have a single method call that basically controls for multiple behaviors that should probably instead be separated. I think there's an opportunity to split these separate functions out through the use of per-platform implementations of a common interface or interfaces. An example with this use case would be:

```typescript
interface PlatformPreview {
  getMobileDevice(
    deviceId?: string,
    logger?: Logger
  ): Promise<IOSSimulatorDevice | AndroidVirtualDevice | undefined>;
}

export class IOSPlatformPreview implements PlatformPreview {
  getMobileDevice(deviceId?: string, logger?: Logger): Promise<any> {
    // Do iOS stuff here.
  }
}

// app.ts
const platformPreview: PlatformPreview = PreviewUtils.platformPreviewFromPlatform(platform);
const device = await platformPreview.getMobileDevice(deviceId, logger);
// etc.
```

If the conditional logic stays as simple as it is now, it's not a big deal. But it has the potential to become challenging to maintain, if it gets more complex.

Reply:
Yes I agree... but I think that should be done in `lwc-dev-mobile-core`, where we already have such separation of logic per platform... and I have a `TODO` comment at the top of this file to move pretty much everything from here to `lwc-dev-mobile-core` once we're almost done with coding the features for the beta release.